### ISSUE TITLE
Extraction generates invalid Haskell code when strings contain unicode characters

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#7870]
#### Version

8.8.0

#### Operating system

Windows (not sure if this matters)

#### Description of the problem

```coq
Require Import Coq.Strings.String.
Require Import Coq.extraction.Extraction.
Require Import Coq.extraction.ExtrHaskellBasic.
Require Import Coq.extraction.ExtrHaskellString.

Definition alpha := "α"%string.
Extraction Language Haskell.
Recursive Extraction alpha.
```
gives
```haskell
module Main where

import qualified Prelude

alpha :: Prelude.String
alpha =
  (:) ' ' ((:) ' ' ([]))


```
where the two characters are rendered in Emacs as `\316` and `\261`.  If I put this in `foo.hs`, then I see
```bash
$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 7.8.3

$ ghc foo.hs -o foo
[1 of 1] Compiling Main             ( foo.hs, foo.o )

foo.hs:7:8:
    lexical error in string/character literal (UTF-8 decoding error)
```

cc @letouzey

**Comment by pi8027** [comment]
@JasonGross Hello. Your script gives me the following output in Coq 8.9 and 8.10.
```
module Main where

import qualified Prelude

alpha :: Prelude.String
alpha =
  (:) '\206' ((:) '\177' ([]))
```
It does compile with GHC 8.0.1. `α` is `0xCE 0xB1` in UTF-8 encoding so I think it's fine. Would you confirm the problem is disappeared? (I have no Windows computer.)

**Comment by pi8027** [comment]
I didn't reproduce this issue in Coq 8.8.0. Pretty-printing of character literals is implemented by using `Char.escaped` and hasn't been changed in the recent 5 years. So I guess that the source of the issue was ocaml/ocaml#6521.

**Comment by llelf** [comment]
> α is 0xCE 0xB1 in UTF-8 encoding so I think it's fine.

Not really.  `"\206\177"` is `"Î±"`

It should be
```haskell
alpha = "α"
   -- or "\945"
   --  = ('\945' : [])
   --  = Data.Text.unpack (Data.Text.Encoding.decodeUtf8 (Data.ByteString.pack [206,177]))
```

**Comment by vbgl** [comment]
Haskell strings (type String) are not encoded: they are a list of unicode code points.

Maybe the extraction of Coq strings should target the ByteString Haskell data type.

**Comment by pi8027** [comment]
@llelf I know. But translating the given Coq definition `alpha` to `"\945"` by Haskell extraction doesn't preserve the meaning. The former one has size 2 but the latter one has size 1.
@vbgl `Data.ByteString.cons` has the linear time complexity so would `[Word8]` be reasonable?

**Comment by vbgl** [comment]
Indeed, a list of Word8 precisely matches the Coq definition. Extracted string literals will probably be hard to read — but they already are…

**Comment by pi8027** [comment]
Anyway, we wouldn't able to deal with Unicode strings in extraction. So I think this issue itself has been addressed in OCaml 4.03.0. But we can keep it opened to find a good alternative to `ExtrHaskellString` which realizes Coq `string` by `[Word8]` or `ByteString`.
